connectors/slack/template/app.js

```diff
@@ -0,0 +1,180 @@
+});
+
+async function registerCore(req, res, isDeregister){
+    if(!req.headers.authorization){
+        res.status(401).send({
+            message: 'Missing auth token'
+        });
+        return;
+    }
+    
+    storage.ensureStorage(req.fusebit);
+
+    await storage.get();
+
+    const [bearerString, token] = req.headers.authorization.split(" ");
+    if(bearerString !== "Bearer" || !Object.keys(storage.tokenToApp).includes(token)){
+        res.status(401).send({
+            message: 'Invalid auth token'
+        });
+        return;
+    }
+
+    let app, handler;
+    if(!isDeregister){
+        if(!req.body || !req.body.app || !req.body.handler){
+            res.status(400).send({
+                message: 'Missing required app and handler body parameters'
+            });
+            return;
+        } else {
+            app = req.body.app;
+            handler = req.body.handler;
+        }
+    } else {
+        if(!req.query || !req.query.app || !req.query.handler){
+             res.status(400).send({
+                message: 'Missing required app and handler query parameters'
+            });
+            return;
+        } else {
+            app = req.query.app;
+            handler = decodeURIComponent(req.query.handler);
+        }
+    }
+
+    const tokenUsedUp = 'TokenUsedUp'; 
+    const registrationProblem = 'RegistrationProblem';
+    
+    try {
+        await storage.put(() => {
+
+            // Doing the token use check in here in case there is a storage conflict
+            // and when the data is refreshed, it turns out the token was used up
+
+            const [storedApp, remainingUses] = storage.tokenToApp[token];
+            if((!isDeregister && storedApp !== app) || remainingUses < 1){
+                Sdk.debug('Token is used up')
+                let e = new Error();
+                e.name = tokenUsedUp;
+                throw e;
+            }
+
+            let handlerRegistered = storage.appToHandler[app] &&
+                storage.appToHandler[app].includes(handler);
+            if((!isDeregister && handlerRegistered) ||
+                (isDeregister && !handlerRegistered)){
+                let e = new Error();
+                e.name = registrationProblem;
+                throw e;
+            }
+
+            if(!isDeregister){
+                Sdk.debug('Registering', handler, app);
+                if(!storage.appToHandler[app]){
+                    storage.appToHandler[app] = [];
+                }
+                storage.appToHandler[app].push(handler);
+            } else {
+                Sdk.debug('Deregistering', handler, app);
+                storage.appToHandler[app].pop(handler);
+                if(storage.appToHandler[app].length == 0){
+                    delete storage.appToHandler[app];
+                }
+            }
+
+            if(remainingUses == 1){
+                Sdk.debug('Token last use')
+                delete storage.tokenToApp[token];
```

Comment:
This doesn't seem necessary.

Reply:
Why so? This cleans up the storage so we're not left with expired tokens